### PR TITLE
Add `cat`, `index_select`, `gather`, and `take_along_axis` opinfo tests.

### PR DESCRIPTION
**Pull request by rdspring1**

### CODE REVIEW

File: python_tests/pytest_input_generators.py [hunk 1]
```diff
@@ -289,6 +345,169 @@ def define_tensor_error_generator(
...
+        b = make_index(shape_b, low=0, high=shape_a[dim])
+        yield SampleInput(a, b, dim)
+
+
+def take_along_axis_error_generator(
+    op: OpInfo, dtype: torch.dtype, requires_grad: bool = False, **kwargs
+):
+    # torch.gather(input: Tensor, dim: int, index: LongTensor)
+    # * input and index tensors have same ndims.
+    # * index tensors must be smaller than input tensor along all dims except specified axis.
```
Comment (jjsjann123) [inline python_tests/pytest_input_generators.py:380]: We have a slightly different protocol here. We stick with numpy.take_along_axis and expect the shapes to be `broadcast` compatible across all dims, except the given axis

File: python_tests/pytest_input_generators.py [hunk 2]
```diff
@@ -289,6 +345,169 @@ def define_tensor_error_generator(
...
+    # * If no dim argument, flatten tensors.
+
+    make_arg = partial(
+        make_tensor, device="cuda", dtype=dtype, requires_grad=requires_grad
+    )
+    make_index = partial(
+        make_tensor, device="cuda", dtype=torch.long, requires_grad=False
+    )
+
+    input_shape = (4, 2)
```
Comment (jjsjann123) [inline python_tests/pytest_input_generators.py:393]: nitpick on code here. It's very hard for me to figure out what combinations are used with this setup. IMHO, some repetition where we have all configuration for a test case grouped together would make it much easier to read

File: python_tests/pytest_input_generators.py [hunk 3]
```diff
@@ -289,6 +345,169 @@ def define_tensor_error_generator(
...
+):
+    make_arg = partial(
+        make_tensor, device="cuda", dtype=dtype, requires_grad=requires_grad
+    )
+    make_index = partial(
+        make_tensor, device="cuda", dtype=torch.long, requires_grad=False
+    )
+
+    # a.shape, dim, b.shape
+    cases = (
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply python_tests/pytest_input_generators.py:359]: I'm actually wondering if we have good support on those tho... :confused: 
maybe we don't need to worry about that for now.

[the remark this answers is not in the document]
Comment (jjsjann123) [reply python_tests/pytest_input_generators.py:359]: Since this is opinfo test, we might as well use this opportunity to make it right. Let's add negative dim to the test case. If it breaks, we can either patch the frontend or open an issue to track it.

**Review by jjsjann123 — APPROVED** [review]
LGTM.

There's a due update that we should have pulled off earlier, `index_select` should be renamed to `take` in nvfuser, since we are generally taking the numpy API rather than pytorch.

File: python_tests/pytest_input_generators.py [hunk 4]
```diff
@@ -289,6 +345,169 @@ def define_tensor_error_generator(
...
+        negative_axis,
+    ]
+
+    for dim, ex_type, ex_str in error_cases:
+        b = make_index(valid_index_shape, low=0, high=10, dtype=torch.long)
+        yield SampleInput(a, b, dim), ex_type, ex_str
+
+    # TODO FIX
+    # b = make_index(valid_index_shape, low=0, high=input_shape[0], dtype=torch.float)
+    # yield SampleInput(a, b, 0), RuntimeError, "index tensor can only be int or long dtype."
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply python_tests/pytest_input_generators.py:415]: These tests are commented out because we don't yield an exception for these cases.

Other frameworks can accept `int` dtype index tensors.

File: python_tests/pytest_input_generators.py [hunk 5]
```diff
@@ -13,6 +13,62 @@
...
+    # shapes, dim
+    cases = [
+        ([(3,)], 0),  # single tensor provided
+        # 1D
+        ([(2,), (3,)], 0),
+        ([(2,), (4,)], 0),
+        ([(0,), (2,)], 0),
+        ([(0,), (2,)], -1),
+        ([(2, 3), (2, 4)], 1),
+        ([(2, 3), (2, 4), (2, 5)], 1),
```
Comment (jacobhinkle) [inline python_tests/pytest_input_generators.py:31]: Could you put a comment here about size one and link to issue https://github.com/NVIDIA/Fuser/issues/224?

File: python_tests/pytest_ops.py [hunk 6]
```diff
@@ -20,6 +20,10 @@ def is_pre_volta():
     return prop.major < 7
 
 
+def is_tensor(a):
+    return type(a) is torch.Tensor
```
Comment (jacobhinkle) [inline python_tests/pytest_ops.py:24]: Note `isinstance(a, torch.Tensor)` will also catch subclasses.

File: python_tests/pytest_ops.py [hunk 7]
```diff
@@ -32,6 +36,8 @@ def parse_inputs_fusion_definition(fd: FusionDefinition, opinfo: OpInfo, *args):
         if is_symbolic:
             if type(a) is torch.Tensor:
                 nvf_args.append(fd.from_pytorch(a))
+            elif type(a) is list and all(map(is_tensor, a)):
```
Comment (jacobhinkle) [inline python_tests/pytest_ops.py:39]: See comment above about `isinstance`

**Review by jacobhinkle — APPROVED** [review]
Looks good. Just a couple nitpicks and comment about cat with broadcast dims.

**Comment by rdspring1** [comment]
!build
